source/memory_pool.c

```diff
@@ -0,0 +1,97 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/memory_pool.h>
+
+int aws_memory_pool_init(struct aws_memory_pool *pool, struct aws_allocator* alloc, size_t element_size, int element_count) {
+    size_t stride = element_size + sizeof(void *);
+    size_t arena_size = stride * element_count;
+    pool->alloc = alloc;
+    pool->arena_size = arena_size;
+    pool->element_size = element_size;
+    pool->arena = alloc->mem_acquire(alloc, arena_size);
+    pool->free_list = pool->arena;
+    pool->overflow_count = 0;
+
+    if (!pool->arena) {
+        aws_raise_error(AWS_ERROR_OOM);
+        return AWS_OP_ERR;
+    }
+
+    /* Hook up singly linked list to represent the free list. */
+    for (int i = 0; i < element_count - 1; ++i) {
+        void **element_ptr = (void **)AWS_PTR_ADD(pool->arena, stride * i);
+        void *next = AWS_PTR_ADD(pool->arena, stride * (i + 1));
```

Comment:
If you just stored `arena` as uint8_t* wouldn't you not have to bother with these macros? That's what we've been doing in other buffer-esque code

Reply:
Yes! Great idea. I'll try it out.